### PR TITLE
feat(text editor): add explicit menu types

### PR DESCRIPTION
**Pull request by john-traas**
Fixes https://github.com/Lundalogik/crm-feature/issues/4084

### CODE REVIEW

File: src/components/text-editor/prosemirror-adapter/menu/menu-commands.ts [hunk 1]
```diff
@@ -1,30 +1,20 @@
...
 import { Schema, MarkType } from 'prosemirror-model';
 import { wrapInList } from 'prosemirror-schema-list';
+import { EditorMenuType } from './types';
 
 export class MenuCommandFactory {
     private schema: Schema;
+    private types: EditorMenuType[];
 
-    constructor(schema: Schema) {
+    constructor(schema: Schema, types: EditorMenuType[]) {
```
[the remark this answers is not in the document]
Comment (john-traas) [reply src/components/text-editor/prosemirror-adapter/menu/menu-commands.ts:10]: I struggled with naming this as well. Prosemirror defines the various options as types, MarkType / NodeType / WrapInType etc. This seemed to make sense.

File: src/components/text-editor/prosemirror-adapter/menu/types.ts [hunk 2]
```diff
@@ -54,3 +54,21 @@ export type EditorPluginButton =
     | 'lift out'
     | 'undo'
     | 'redo';
+
+/**
+ * The `EditorMenuType` type is used to specify the type of menu items that can be added to the editor toolbar.
+ * Each string represents a different type to be used for creating the prosemirror commands relevate to the button.
```
[the remark this answers is not in the document]
Comment (john-traas) [reply src/components/text-editor/prosemirror-adapter/menu/types.ts]: Fixed in fixup.

**Review by LawrenceBorst — APPROVED** [review]

File: src/components/text-editor/prosemirror-adapter/menu/menu-commands.ts [hunk 3]
```diff
@@ -1,93 +1,115 @@
...
-        return wrapIn(type);
+    constructor(schema: Schema) {
+        this.schema = schema;
     }
 
-    private createListCommand(listType: string) {
-        const type = this.schema.nodes[listType];
-        if (!type) {
-            throw new Error(`List type "${listType}" not found in schema`);
+    getCommand(mark: EditorMenuTypes, url?: string) {
```
Comment (civing) [inline src/components/text-editor/prosemirror-adapter/menu/menu-commands.ts:93]: Would be nice to specify this function as `public`
```suggestion
    public getCommand(mark: EditorMenuTypes, url?: string) {
```

File: src/components/text-editor/prosemirror-adapter/menu/menu-commands.ts [hunk 4]
```diff
@@ -1,93 +1,115 @@
...
+        if (!Object.values(EditorMenuTypes).includes(mark)) {
+            throw new Error(`The Menu Type "${mark}" is not supported`);
+        }
```
Comment (civing) [inline src/components/text-editor/prosemirror-adapter/menu/menu-commands.ts:96]: Is this really necessary? Can't we trust the consumers of `getCommand` to do type checking?

File: src/components/text-editor/prosemirror-adapter/menu/types.ts [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+    Strong = 'strong',
+    Italic = 'em',
+    Blockquote = 'blockquote',
+    HeaderLevel1 = 'headerlevel1',
+    HeaderLevel2 = 'headerlevel2',
+    HeaderLevel3 = 'headerlevel3',
+    Link = 'link',
+    OrderedList = 'ordered_list',
+    BulletList = 'bullet_list',
+}
```
Comment (civing) [inline src/components/text-editor/prosemirror-adapter/menu/types.ts:18]: What about using a string literal union type instead?

```suggestion
export type EditorMenuTypes =
    | 'strong'
    | 'em'
    | 'blockquote'
    | 'headerlevel1'
    | 'headerlevel2'
    | 'headerlevel3'
    | 'link'
    | 'ordered_list'
    | 'bullet_list';
```

File: src/components/text-editor/prosemirror-adapter/menu/menu-commands.ts (hunk 3 above)
Comment (john-traas) [reply src/components/text-editor/prosemirror-adapter/menu/menu-commands.ts:93]: Ah yes, I agree.

File: src/components/text-editor/prosemirror-adapter/menu/menu-commands.ts (hunk 4 above)
Comment (john-traas) [reply src/components/text-editor/prosemirror-adapter/menu/menu-commands.ts:96]: We could definitely do that.

File: src/components/text-editor/prosemirror-adapter/prosemirror-adapter.tsx [hunk 6]
```diff
@@ -198,6 +211,12 @@ export class ProsemirrorAdapter {
         });
         this.view.dispatch(transaction);
 
-        this.view.focus();
+        this.setFocus();
+    }
+
+    public setFocus() {
+        if (this.view) {
+            this.view.focus();
```
Comment (LawrenceBorst) [inline src/components/text-editor/prosemirror-adapter/prosemirror-adapter.tsx:219]: We could also use `this.view?.focus()` to keep it short

**Review by LawrenceBorst — APPROVED** [review]

File: src/components/text-editor/prosemirror-adapter/menu/menu-commands.ts [hunk 7]
```diff
@@ -1,93 +1,115 @@
...
 import { toggleMark, setBlockType, wrapIn } from 'prosemirror-commands';
-import { Schema, MarkType } from 'prosemirror-model';
+import { Schema, MarkType, NodeType } from 'prosemirror-model';
 import { wrapInList } from 'prosemirror-schema-list';
+import { Command } from 'prosemirror-state';
+import { EditorMenuTypes, levelMapping } from './types';
 
-export class MenuCommandFactory {
-    private schema: Schema;
+type CommandFunction = (
```
[the remark this answers is not in the document]
Comment (john-traas) [reply src/components/text-editor/prosemirror-adapter/menu/menu-commands.ts:8]: My thinking with explicitly naming it `Menu` CommandFactory was so as not to confuse it with anything related to platform commands. Even though this is within Lime Elements I still felt that the distinction kept things clearer.

File: src/components/text-editor/prosemirror-adapter/prosemirror-adapter.tsx (hunk 6 above)
Comment (john-traas) [reply src/components/text-editor/prosemirror-adapter/prosemirror-adapter.tsx]: Updated :)

**Comment by john-traas** [comment]
I've simplified the current tests as I ran into some issues with mocking the editor state, dom and global selection in the tests. I'm working on it but I have a follow-up PR that refactors the prosemirror-adaptor to streamline it a bit more and I think we can add to the tests in that PR if need be.

**Review by civing — DISMISSED** [review]
I think this looks really clean and tidy now. It would be great to fix @LawrenceBorst comment about not disabling camel case lint check for the entire file, otherwise I think you can merge this.

File: src/components/text-editor/prosemirror-adapter/menu/menu-commands.ts [hunk 8]
```diff
@@ -1,93 +1,115 @@
+/* eslint-disable camelcase */
```
[the remark this answers is not in the document]
Comment (john-traas) [reply src/components/text-editor/prosemirror-adapter/menu/menu-commands.ts]: Updated. Can go ahead and re-approve this one @LawrenceBorst 😄

**Review by LawrenceBorst — APPROVED** [review]

**Pull request merged**